### ISSUE TITLE
[Question]: Training Checkpointing and Resumption for llama.cpp Fine-tuning

### ISSUE THREAD

**Issue by makaveli10** [ggml-org/llama.cpp#15442]
This RFC proposes adding checkpointing and resumption capabilities to fine-tuning in llama.cpp (examples/finetune.cpp). The implementation would enable:
 - Step-based checkpointing during training
 - model weights + optimizer state + metadata(epoch+step..)
 - Training resumption from exact same point

**Motivation**

Currently, fine-tuning in llama.cpp lacks checkpointing capabilities, making it vulnerable to:
 - Training loss due to hardware failures or interruptions
 - Resource waste when restarting long training runs
 - Limited experimentation ability (can't resume from different points)

----------------------------------------------------------------------------------------------
**Core Technical Challenge: Optimizer State Loading**

Optimizer tensors don't exist until after ggml_opt_alloc() is called during the first training step. This creates a timing challenge:

  1. Initialization: We know we want to load optimizer state from a checkpoint
  2. First allocation: ggml_opt_alloc() creates optimizer tensors (momentum, variance arrays)
  3. Loading opportunity: Only after allocation can we restore saved optimizer state

Current Training Flow
```
  // examples/finetune.cpp current flow:
  llama_opt_init(ctx.get(), model.get(), lopt_params);
  for (int epoch = 0; epoch < 2; ++epoch) {
      llama_opt_epoch(ctx.get(), dataset, result_train, result_eval, idata_split,
          ggml_opt_epoch_callback_progress_bar, ggml_opt_epoch_callback_progress_bar, -1);
  }
```

Problem: No hook exists between ggml_opt_alloc() and optimizer step execution.

---------------------------------------------------------------------------------

**_Proposed Solutions_**

```
// Enhanced llama_opt_params

  struct llama_opt_params {
      // ... existing fields ...
      const char* checkpoint_path;        // Path to checkpoint file
      bool load_optimizer_state;          // Enable optimizer state loading
  };

// Enhanced llama_opt_epoch with Resume Support

  void llama_opt_epoch(
      llama_context* ctx,
      ggml_opt_dataset_t dataset,
      ggml_opt_result_t result_train,
      ggml_opt_result_t result_eval,
      int64_t idata_split,
      ggml_opt_epoch_callback callback_train,
      ggml_opt_epoch_callback callback_eval,
      int64_t resume_from_batch = -1      // Resume from specific batch
  );

// Implementation Flow

  // 1. Initialize with checkpoint info
  struct llama_opt_params lopt_params {
      // ... existing fields ...
      .checkpoint_path = "checkpoint_step_00001200.gguf",
      .load_optimizer_state = true,
  };
  llama_opt_init(ctx.get(), model.get(), lopt_params);

  // 2. Start training with resume point
  llama_opt_epoch(ctx.get(), dataset, result_train, result_eval, idata_split,
      callback_train, callback_eval, resume_batch);
```

**Delayed Loading**
Inside the first training step: After ggml_opt_alloc() creates tensors, but before optimization begins, we automatically load the saved optimizer state.

- Timeline:
  1. llama_opt_init() - Store checkpoint path for later
  2. llama_opt_epoch() - Begin epoch processing
  3. First batch processing -> ggml_opt_alloc() creates tensors
  4. Automatic loading - Restore optimizer state from checkpoint
  5. Continue training from exact point

- New APIs Required

  - Optimizer State Access
  ```
  // Access optimizer internal state for saving/loading
  int32_t llama_opt_get_nparams(llama_context* ctx);
  struct ggml_tensor* llama_opt_get_grad_m(llama_context* ctx, int32_t index);
  struct ggml_tensor* llama_opt_get_grad_v(llama_context* ctx, int32_t index);
  int64_t llama_opt_get_iter(llama_context* ctx);
  void llama_opt_set_iter(llama_context* ctx, int64_t iter);
  ```
  - Checkpoint Management
  ```
  // Save complete training state
  bool llama_save_training_checkpoint(
      llama_context* ctx,
      const char* checkpoint_path,
      int64_t current_step,
      int32_t current_epoch
  );

  // Load optimizer state (called automatically)
  bool llama_load_optimizer_state_from_checkpoint(
      llama_context* ctx, 
      const char* checkpoint_path
  );
  ```

**Key Questions for Community**

  1. API Integration Approach

  - Should we extend existing `llama_opt_params` and `llama_opt_epoch()` as proposed?
  - Or create entirely separate checkpoint APIs?
  - Is automatic optimizer state loading after `ggml_opt_alloc()` the right approach? 
  - Which would happen only once based on stored checkpoint info in `llama_context`:
    ```
    // Added to llama_context
    std::string pending_optimizer_checkpoint;
    bool optimizer_state_loaded = false;
    ```

  2. Optimizer State Format

  - Raw arrays: Direct serialization of momentum/variance data (my current approach)?
  - GGUF integration: Store optimizer tensors in extended GGUF format?
  - Structured binary: Custom format with metadata?

  3. File Organization

  - My current approach uses 3 files per checkpoint:
  ```
  checkpoint_step_00001200.gguf      # Model weights
  checkpoint_step_00001200.optstate  # Optimizer state (binary)
  checkpoint_step_00001200.meta      # Training metadata (text)
  ```
  - Is this multi-file approach preferred?
  - Or should we embed everything in GGUF?
  - Or use a directory-based approach?

  4. Resumption Granularity

  - Batch-level resumption: My proposed implementation would resume from the exact next unseen batch
  - Is this the right granularity, or should we support epoch-level only?
  - How important is step-perfect resumption vs approximate resumption?

**Seeking Community Input**

I'm looking for guidance on the architectural decisions outlined above before proceeding with a full implementation. The main questions are around API design, file format preferences, and integration strategy.

Would the community be interested in this feature? Are there alternative approaches I should consider?

**Comment by JohannesGaessler** [comment]
In `llama-model-saver.cpp` there is already code for converting a `llama_model` to a `gguf_ctx` and saving the model to a file. The API in `ggml-opt.h` should be extended with 2 functions like `ggml_opt_save_state` and `ggml_opt_load_state` that save and load the current `ggml_opt` state to a GGUF file. The checkpointing logic should be in `finetune.cpp`, long-term it may make sense to move it. The model and the `ggml_opt` state after a certain epoch should be saved as separate files to the same directory.

**Comment by makaveli10** [comment]
@JohannesGaessler Thank you for the clear architectural guidance.

I'll implement this approach:
  - Add `ggml_opt_save_state()` and `ggml_opt_load_state()` to ggml-opt.h using GGUF format
  - Use existing model saver patterns from llama-model-saver.cpp
  - Two-file structure: checkpoint_dir/model.gguf + checkpoint_dir/optimizer.gguf
  - Keep checkpointing orchestration in finetune.cpp

**_Quick clarification on timing_**: Since optimizer tensors don't exist until after `ggml_opt_alloc()` (during first training step), I'll still need the delayed loading mechanism(?) where `ggml_opt_load_state()` gets called automatically after allocation but before optimization begins. The new API will just encapsulate the state loading logic cleanly.

I'll work on this implementation and submit a PR following this architecture. Thanks again!

**Comment by JohannesGaessler** [comment]
The function `ggml_opt_load_state` should initialize `ctx_static`, `grad_m`, and `grad_v`. The function `ggml_opt_build` already has checks to initialize these only once so there shouldn't be issues regarding timing.

I'm not 100% sure whether `grad_accs`, the tensors for gradient accumulation, should also be saved and loaded. I think they should since I think it will make the implementation simpler and more flexible.

**Comment by makaveli10** [comment]
@JohannesGaessler Thanks for the swift response. So, the only place I see I can load the saved optimizer state is 
https://github.com/ggml-org/llama.cpp/blob/657b8a77bd01854f99d37a47318fa24f2e7e298f/src/llama-context.cpp#L2158

```
  struct ggml_context * ctx_compute_opt;
  {
      const size_t size_gf = ggml_graph_size(gf);
      const size_t size_meta = 4*size_gf*ggml_tensor_overhead() + 2*ggml_graph_overhead_custom(size_gf, /*grads = */ true);
      struct ggml_init_params params = {
          /*.mem_size   =*/ size_meta,
          /*.mem_buffer =*/ nullptr,
          /*.no_alloc   =*/ true,
      };
      ctx_compute_opt = ggml_init(params);
  }
  ctx_compute_opt = ggml_init(params) // creates the compute context
  ggml_opt_prepare_alloc(opt_ctx, ctx_compute_opt, gf, ...) // sets it on the optimizer context
  ggml_opt_alloc(opt_ctx, train) // allocates the optimizer tensors

```
And only after `ggml_opt_alloc` I can actually load the saved `grad_v` & `grad_m`.
Where we load it something like this:
```
opt_iter(){
    ....
    ....
    ggml_opt_prepare_alloc(opt_ctx, ctx_compute_opt, gf, res->get_tokens(), res->get_logits());
    ggml_opt_alloc(opt_ctx, train);
    
    // Load optimizer state once from checkpoint after first successful allocation
    if (!optimizer_state_loaded && !pending_optimizer_checkpoint.empty()) {
        std::filesystem::path checkpoint_path(pending_optimizer_checkpoint);
        std::filesystem::path checkpoint_dir = checkpoint_path.parent_path();
        std::filesystem::path optimizer_path = checkpoint_dir / "optimizer.gguf";

        // opt_load_state calls ggml_opt_load_state() which loads the saved grad_m, grad_v
        if (opt_load_state(optimizer_path.c_str())) {
            LLAMA_LOG_INFO("Optimizer state loaded from checkpoint: %s\n", optimizer_path.c_str());
        } else {
            LLAMA_LOG_WARN("Failed to load optimizer state from checkpoint: %s\n", optimizer_path.c_str());
        }
        optimizer_state_loaded = true;  // Only try once
        pending_optimizer_checkpoint.clear();
    }
}
```
Does this make sense?

**Comment by JohannesGaessler** [comment]
There should be no need to call or modify `ggml_opt_alloc` or `ggml_opt_build` for loading the state of `ggml_opt_context`. All you need to do is save/load `ctx_static` with the pre-existing GGUF code and implement code for saving/loading `grad_accs`, `grad_m`, `grad_v`.
